services/server/src/handlers/availabilitySlot.ts

```diff
@@ -0,0 +1,235 @@
+import { datetime, id, pagination } from "@/validation/utils";
+import { isTutor, isTutorManager, isUser } from "@litespace/auth";
+import { IAvailabilitySlot } from "@litespace/types";
+import {
+  availabilitySlots,
+  interviews,
+  knex,
+  lessons,
+} from "@litespace/models";
+import dayjs from "@/lib/dayjs";
+import { NextFunction, Request, Response } from "express";
+import safeRequest from "express-async-handler";
+import { INTERVIEW_DURATION, isIntersecting } from "@litespace/sol";
+import zod from "zod";
+import { isEmpty } from "lodash";
+
+const findPayload = zod.object({
+  userId: id,
+  after: datetime,
+  before: datetime,
+  pagination,
+});
+
+const setPayload = zod.object({
+  slots: zod.array(
+    zod
+      .object({
+        action: zod.literal("create"),
+        start: datetime,
+        end: datetime,
+      })
+      .or(
+        zod
+          .object({
+            action: zod.literal("update"),
+            id: id,
+            start: datetime.optional(),
+            end: datetime.optional(),
+          })
+          .or(
+            zod.object({
+              action: zod.literal("delete"),
+              id: id,
+            })
+          )
+      )
+  ),
+});
+
+type CreateAction = {
+  action: "create";
+  start: string;
+  end: string;
+};
+type UpdateAction = {
+  action: "update";
+  id: number;
+  start?: string;
+  end?: string;
+};
+type DeleteAction = {
+  action: "delete";
+  id: number;
+};
+
+async function find(req: Request, res: Response, next: NextFunction) {
+  const user = req.user;
+  if (!isUser(user)) return next(forbidden());
+
+  const { userId, after, before, pagination } = findPayload.parse(req.body);
+
+  const diff = dayjs(before).diff(after, "days");
+  if (diff < 0) return next(bad());
+
+  pagination.full = pagination.full && diff <= 30;
+  const paginatedSlots = await availabilitySlots.find({
+    users: [userId],
+    after,
+    before,
+    pagination,
+  });
+
+  const slotIds = paginatedSlots.list.map((slot) => slot.id);
+
+  const paginatedLessons = await lessons.find({
+    users: [userId],
+    slots: slotIds,
+    after,
+    before,
+  });
+
+  const paginatedInterviews = await interviews.find({
+    users: [userId],
+    slots: slotIds,
+  });
+
+  const result: IAvailabilitySlot.FindAvailabilitySlotsApiResponse = {
+    list: [
+      {
+        slots: paginatedSlots.list,
+        subslots: [
+          ...paginatedLessons.list.map((lesson) => ({
+            parent: lesson.slotId,
+            start: lesson.start,
+            end: dayjs(lesson.start)
+              .add(lesson.duration, "minutes")
+              .toISOString(),
+          })),
+          ...paginatedInterviews.list.map((interview) => ({
+            parent: interview.ids.slot,
+            start: interview.start,
+            end: dayjs(interview.start)
+              .add(INTERVIEW_DURATION, "minutes")
+              .toISOString(),
+          })),
```

Comment:
We can make the logic more clean by creating a function similar to `asSlot`. You can find it [here](https://github.com/litespace-org/litespace/blob/aebee0e646360ee861098e57b12dd029098fca1a/packages/sol/src/rule.ts#L493-L507).


> [!NOTE]
> `asSlot` and `asSlots` are located in `rules.ts` file which will be remove very soon. You can copy them (or move) them to `sol/src/availiabiltySlots.ts` file.